cpp/monitoring/gcm/exporter.cc

```diff
@@ -79,29 +84,35 @@ GCMExporter::~GCMExporter() {
 
 
 void GCMExporter::RefreshCredentials() {
-  VLOG(1) << "Refreshing GCM credentials...";
-  UrlFetcher::Request req(
-      (URL(FLAGS_google_compute_metadata_url + "/" +
-           FLAGS_google_compute_monitoring_service_account + "/token")));
-  req.headers.insert(make_pair("Metadata-Flavor", "Google"));
+  do {
+    VLOG(1) << "Refreshing GCM credentials...";
+    UrlFetcher::Request req(
+        (URL(FLAGS_google_compute_metadata_url + "/" +
+             FLAGS_google_compute_monitoring_service_account + "/token")));
+    req.headers.insert(make_pair("Metadata-Flavor", "Google"));
 
-  UrlFetcher::Response resp;
-  SyncTask task(executor_);
-  fetcher_->Fetch(req, &resp, task.task());
-  task.Wait();
+    UrlFetcher::Response resp;
+    SyncTask task(executor_);
+    fetcher_->Fetch(req, &resp, task.task());
+    task.Wait();
 
-  // TODO(alcutter): Handle this better
-  CHECK_EQ(util::Status::OK, task.status());
-  CHECK_EQ(200, resp.status_code);
-  token_refreshed_at_ = system_clock::now();
+    if (!task.status().ok() || resp.status_code != 200) {
+      LOG(WARNING) << "Failed to refresh GCM credentials, status: "
+                   << task.status() << ", response code: " << resp.status_code;
+      sleep(FLAGS_google_compute_monitoring_retry_delay_seconds);
```

Comment:
All this stuff runs on some thread pool, I'm presuming, but it's not of unlimited size... "One more thing" isn't going to stuff us right there and then, but seeing the `SyncTask` above made me think it could be split up (for example, `RefreshCredentials` taking a `util::Task`, doing its thing async with child tasks, and `PushMetrics` being split at the point it calls `RefreshCredentials`, either setting the second part as the callback for the child task, or calling the second part directly).

Could be a separate PR, but think about it? Maybe file an issue, to remember?